packages/yarnpkg-core/sources/Configuration.ts

```diff
@@ -1126,23 +1143,40 @@ export class Configuration {
         if (!Array.isArray(data.plugins))
           continue;
 
+        const missingPluginMetaList = [];
         for (const userPluginEntry of data.plugins) {
           const userProvidedPath = typeof userPluginEntry !== `string`
             ? userPluginEntry.path
             : userPluginEntry;
+          const userProvidedSpec = userPluginEntry?.spec ?? ``;
 
           const pluginPath = ppath.resolve(cwd, npath.toPortablePath(userProvidedPath));
+          if (!await xfs.existsPromise(pluginPath)) {
+            if (!userProvidedSpec.match(/^https?:/)) {
+              const prettyPluginName = formatUtils.pretty(configuration, ppath.basename(pluginPath, `.cjs`), formatUtils.Type.NAME);
+              const prettyGitIgnore = formatUtils.pretty(configuration, `.gitignore`, formatUtils.Type.NAME) ;
+              const prettyYarnrc = formatUtils.pretty(configuration, configuration.values.get(`rcFilename`), formatUtils.Type.NAME) ;
+              const prettyUrl = formatUtils.pretty(configuration, `https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored`, formatUtils.Type.URL) ;
+              throw new UsageError(`Failed to get ${prettyPluginName} plugin, please try to delete the plugin from ${prettyYarnrc} and reinstall it, the error usually occurs because ${prettyGitIgnore} is incorrect, please check ${prettyUrl} to set the correct ${prettyGitIgnore}`);
+            }
+
+            const pluginBuffer = await httpUtils.get(userProvidedSpec, {configuration});
+            await xfs.mkdirPromise(ppath.dirname(pluginPath), {recursive: true});
+            await xfs.writeFilePromise(pluginPath, pluginBuffer);
+
+            missingPluginMetaList.push({
+              path: userProvidedPath,
+              spec: userProvidedSpec,
+              checksum: await hashUtils.checksumFile(pluginPath),
+            });
+          }
           await importPlugin(pluginPath, path);
         }
+        await Configuration.addPlugin(cwd, missingPluginMetaList);
```

Comment:
Why do we need to call `addPlugin` here (which only mutates the rc file if I understand it right), since we're just loading missing plugins that are already part of the rc files?

Reply:
even if the plugin was downloaded from the same URL, we still cannot guarantee that it has not been modified.

And I'm guessing that the reason the plugin is missing is probably because it didn't add `!.yarn/plugin` to `.gitignore` 🤔?

```yml
.pnp.*
.yarn/*
!.yarn/patches
!.yarn/releases
!.yarn/sdks
!.yarn/versions
# !.yarn/plugins is missing
```

So if the plugin file at that URL is modified, the user will only be able to observe the change from `yarnrc.yml`, just like `yarn.lock`.

Reply:
See other comment - if we refuse to install plugins that don't match the checksum, we don't have this problem anymore. Should users want to update their plugins, they can run a `yarn plugin import` again (although it might still crash because the plugin will be missing and would fail the checksum check, but that's something we can figure out if we need to - we could also say that plugins must have unique urls, as a rule).

Reply:
I'm deleting it here 

https://github.com/yarnpkg/berry/pull/4912/commits/41c3a6e47a8a410978d437e403876939bb4e599a